pkg/ccl/changefeedccl/sink_kafka_v2.go

```diff
@@ -0,0 +1,603 @@
+	// NOTE: kgo lets you give multiple compression options in preference order, which is cool but the config json doesnt support that. Should we?
+	switch sarama.CompressionCodec(sinkCfg.Compression) {
+	case sarama.CompressionNone:
+		opts = append(opts, kgo.ProducerBatchCompression(kgo.NoCompression()))
+	case sarama.CompressionGZIP:
+		opts = append(opts, kgo.ProducerBatchCompression(kgo.GzipCompression()))
+	case sarama.CompressionSnappy:
+		opts = append(opts, kgo.ProducerBatchCompression(kgo.SnappyCompression()))
+	case sarama.CompressionLZ4:
+		opts = append(opts, kgo.ProducerBatchCompression(kgo.Lz4Compression()))
+	case sarama.CompressionZSTD:
+		opts = append(opts, kgo.ProducerBatchCompression(kgo.ZstdCompression()))
+	default:
+		return nil, errors.Errorf(`unknown compression codec: %v`, sinkCfg.Compression)
+	}
+
+	if version := sinkCfg.Version; version != "" {
+		if !strings.HasPrefix(version, `v`) {
+			version = `v` + version
+		}
+		v := kversion.FromString(version)
+		if v == nil {
+			return nil, errors.Errorf(`unknown kafka version: %s`, version)
+		}
+		// NOTE: This version of kgo doesnt support specifying max versions
+		// >3.6.0 (released Oct 10 2023). This option is only really needed for
+		// interacting with <v0.10.0 clusters anyway.
+		opts = append(opts, kgo.MaxVersions(v))
+	}
+
+	return opts, nil
+}
+
+type kgoLogAdapter struct {
+	ctx context.Context
+}
+
+func (k kgoLogAdapter) Level() kgo.LogLevel {
+	if log.V(2) {
+		return kgo.LogLevelDebug
+	}
+	return kgo.LogLevelInfo
+}
+
+func (k kgoLogAdapter) Log(level kgo.LogLevel, msg string, keyvals ...any) {
+	format := `kafka(kgo): %s %s`
+	for i := 0; i < len(keyvals); i += 2 {
+		format += ` %s=%v`
+	}
+	log.InfofDepth(k.ctx, 1, format, append([]any{redact.SafeString(level.String()), redact.SafeString(msg)}, keyvals...)...) //nolint:fmtsafe
+}
+
+var _ kgo.Logger = kgoLogAdapter{}
+
+// wrappers around the recommended sarama compat approach to pass thru record partitions when key is nil, like current sarama impl
+func newKgoChangefeedPartitioner() kgo.Partitioner {
+	hasher := fnv.New32a()
+	return &kgoChangefeedPartitioner{
+		inner: kgo.StickyKeyPartitioner(kgo.SaramaCompatHasher(func(bs []byte) uint32 {
+			hasher.Reset()
+			_, _ = hasher.Write(bs)
+			return hasher.Sum32()
+		})),
+	}
+}
+
+type kgoChangefeedPartitioner struct {
+	inner kgo.Partitioner
+}
+
+func (p *kgoChangefeedPartitioner) ForTopic(topic string) kgo.TopicPartitioner {
+	return &kgoChangefeedTopicPartitioner{inner: p.inner.ForTopic(topic)}
+}
+
+type kgoChangefeedTopicPartitioner struct {
+	inner kgo.TopicPartitioner
+}
+
+func (p *kgoChangefeedTopicPartitioner) RequiresConsistency(*kgo.Record) bool { return true }
+func (p *kgoChangefeedTopicPartitioner) Partition(r *kgo.Record, n int) int {
+	// Let messages without keys specify where they want to go (for resolved messages). Note that the CSV encoder always provides nil keys, so they will always go to partition 0.
```

Comment:
this should be fixed by the bit in batchbuffer.append that sets key = []byte{}. TODO: document this and make a test for it